lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp

```diff
@@ -0,0 +1,809 @@
+    ThunkBody->createFunctionArgument(ArgDesc.Arg->getType(), ArgDesc.Decl);
+  }
+
+  /// Builder to add new instructions in the Thunk.
+  SILBuilder Builder(ThunkBody);
+  SILOpenedArchetypesTracker OpenedArchetypesTracker(F);
+  Builder.setOpenedArchetypesTracker(&OpenedArchetypesTracker);
+  Builder.setCurrentDebugScope(ThunkBody->getParent()->getDebugScope());
+
+  /// Location to insert new instructions.
+  SILLocation Loc = ThunkBody->getParent()->getLocation();
+
+  /// Create the function_ref instruction to the NewF.
+  FunctionRefInst *FRI = Builder.createFunctionRef(Loc, NewF);
+
+  auto GenCalleeType = NewF->getLoweredFunctionType();
+  auto CalleeGenericSig = GenCalleeType->getGenericSignature();
+
+  /// Determine arguments to Apply.
+  /// Generate opened existentials for generics.
+  llvm::SmallVector<SILValue, 8> ApplyArgs;
+
+  /// Ideally this should be a SubstitutionMap. TODO.
+  llvm::DenseMap<SubstitutableType *, Type> Generic2OpenedTypeMap;
+  for (auto &ArgDesc : ArgumentDescList) {
+    if (Arg2GenericTypeMap[ArgDesc.Index]) {
+      ArchetypeType *Opened;
+      SILValue OrigOperand = ThunkBody->getArgument(ArgDesc.Index);
+      auto SwiftType = ArgDesc.Arg->getType().getASTType();
+      auto OpenedType =
+          SwiftType->openAnyExistentialType(Opened)->getCanonicalType();
+      auto OpenedSILType = NewF->getModule().Types.getLoweredType(OpenedType);
+      SILValue archetypeValue;
+      auto ExistentialRepr =
+          ArgDesc.Arg->getType().getPreferredExistentialRepresentation(M);
+      switch (ExistentialRepr) {
+      case ExistentialRepresentation::Opaque: {
+        archetypeValue = Builder.createOpenExistentialAddr(
+            Loc, OrigOperand, OpenedSILType, OpenedExistentialAccess::Mutable);
+        ApplyArgs.push_back(archetypeValue);
+        break;
+      }
+      case ExistentialRepresentation::Class: {
+        archetypeValue =
+            Builder.createOpenExistentialRef(Loc, OrigOperand, OpenedSILType);
+        ApplyArgs.push_back(archetypeValue);
+        break;
+      }
+      default: {
+        assert(true);
+        break;
+      }
+      };
+      auto GenericParam = Arg2GenericTypeMap[ArgDesc.Index];
+      Generic2OpenedTypeMap[GenericParam] = OpenedType;
+    } else {
+      auto CallerArg = ThunkBody->getArgument(ArgDesc.Index);
+      auto SwiftType = CallerArg->getType().getASTType();
+      if (auto *GP = SwiftType->getAs<GenericTypeParamType>()) {
+        ArchetypeType *Opened;
+        auto OpenedType =
+            SwiftType->openAnyExistentialType(Opened)->getCanonicalType();
+        Generic2OpenedTypeMap[GP] = OpenedType;
+      }
+      ApplyArgs.push_back(ThunkBody->getArgument(ArgDesc.Index));
+    }
+  }
+
+  /// Create substitutions for Apply instructions.
+  auto SubMap =
+      SubstitutionMap::get(CalleeGenericSig,
+                           [&](SubstitutableType *type) -> Type {
+                             return Generic2OpenedTypeMap[type];
```

Comment:
This is where you check: is 'type' a GenericTypeParamType that you added, in which case you look up into Generic2OpenedTypeMap, otherwise, you `F->mapTypeIntoContext(type)`.